From 69408e09b0bbb5d795a30daef1d6c2e7254b34d8 Mon Sep 17 00:00:00 2001
From: Ross Wille <wille@freescale.com>
Date: Mon, 16 Feb 2009 08:45:02 -0700
Subject: [PATCH] ENGR00107729-2 MX25: Add FlexCAN support

Add support to MX25 3-stack for the FlexCAN driver.
i.MX25 3-stack has one CAN port that is routed to
the CAN2 interface on the chip.  The CAN transceiver
is enabled by a GPIO on the MX25.

Signed-off-by: Ross Wille <wille@freescale.com>
---
 arch/arm/configs/imx25_3stack_defconfig |   14 ++++++-
 arch/arm/mach-mx25/Kconfig              |    6 +++
 arch/arm/mach-mx25/board-mx25_3stack.h  |    3 ++
 arch/arm/mach-mx25/devices.c            |   60 +++++++++++++++++++++++++++++++
 arch/arm/mach-mx25/mx25_3stack.c        |   33 +++++++++++++++++
 arch/arm/mach-mx25/mx25_3stack_gpio.c   |   41 ++++++++-------------
 include/asm-arm/arch-mxc/mx25.h         |    2 +-
 7 files changed, 130 insertions(+), 29 deletions(-)

diff --git a/arch/arm/configs/imx25_3stack_defconfig b/arch/arm/configs/imx25_3stack_defconfig
index fbadbbf..61f055a 100644
--- a/arch/arm/configs/imx25_3stack_defconfig
+++ b/arch/arm/configs/imx25_3stack_defconfig
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 2.6.26
-# Fri Jan 30 20:51:13 2009
+# Wed Mar 11 16:54:31 2009
 #
 CONFIG_ARM=y
 CONFIG_SYS_SUPPORTS_APM_EMULATION=y
@@ -192,6 +192,7 @@ CONFIG_ARCH_MXC_HAS_NFC_V2_1=y
 # Device options
 #
 # CONFIG_I2C_MXC_SELECT3 is not set
+# CONFIG_FLEXCAN_MXC_SELECT1 is not set
 CONFIG_DMA_ZONE_SIZE=24
 CONFIG_UTMI_MXC=y
 CONFIG_UTMI_MXC_OTG=m
@@ -366,7 +367,16 @@ CONFIG_DEFAULT_TCP_CONG="cubic"
 #
 # CONFIG_NET_PKTGEN is not set
 # CONFIG_HAMRADIO is not set
-# CONFIG_CAN is not set
+CONFIG_CAN=y
+CONFIG_CAN_RAW=y
+# CONFIG_CAN_BCM is not set
+
+#
+# CAN Device Drivers
+#
+# CONFIG_CAN_VCAN is not set
+# CONFIG_CAN_DEBUG_DEVICES is not set
+CONFIG_CAN_FLEXCAN=y
 CONFIG_IRDA=m
 
 #
diff --git a/arch/arm/mach-mx25/Kconfig b/arch/arm/mach-mx25/Kconfig
index 03d27eb..6454987 100644
--- a/arch/arm/mach-mx25/Kconfig
+++ b/arch/arm/mach-mx25/Kconfig
@@ -67,6 +67,12 @@ config I2C_MXC_SELECT3
 	help
 	  Enable MX25 I2C3 module.
 
+config FLEXCAN_MXC_SELECT1
+	bool "Enable FlexCAN1 module"
+	depends on CAN_FLEXCAN
+	help
+	  Enable MX25 FlexCAN1 module.
+
 endmenu
 
 endmenu
diff --git a/arch/arm/mach-mx25/board-mx25_3stack.h b/arch/arm/mach-mx25/board-mx25_3stack.h
index 0c45418..125876b 100644
--- a/arch/arm/mach-mx25/board-mx25_3stack.h
+++ b/arch/arm/mach-mx25/board-mx25_3stack.h
@@ -167,6 +167,9 @@ extern int headphone_det_status(void);
 extern void sgtl5000_enable_amp(void);
 extern unsigned int sdhc_get_card_det_status(struct device *dev);
 extern int sdhc_write_protect(struct device *dev);
+extern void gpio_can_active(int id);
+extern void gpio_can_inactive(int id);
+extern struct flexcan_platform_data flexcan_data[];
 
 #endif				/* CONFIG_MACH_MX25_3DS */
 #endif				/* __ASM_ARCH_MXC_BOARD_MX25_3STACK_H__ */
diff --git a/arch/arm/mach-mx25/devices.c b/arch/arm/mach-mx25/devices.c
index 45e1095..5b7713b 100644
--- a/arch/arm/mach-mx25/devices.c
+++ b/arch/arm/mach-mx25/devices.c
@@ -24,6 +24,7 @@
 
 #include "iomux.h"
 #include "sdma_script_code.h"
+#include "board-mx25_3stack.h"
 
 void mxc_sdma_get_script_info(sdma_script_start_addrs * sdma_script_addr)
 {
@@ -483,6 +484,64 @@ static void imx_init_adc(void)
 }
 #endif
 
+#if defined(CONFIG_CAN_FLEXCAN) || defined(CONFIG_CAN_FLEXCAN_MODULE)
+
+static struct resource flexcan1_resources[] = {
+	{
+	 .start = CAN1_BASE_ADDR,
+	 .end = CAN1_BASE_ADDR + 0x97F,
+	 .flags = IORESOURCE_MEM,},
+	{
+	 .start = MXC_INT_CAN1,
+	 .end = MXC_INT_CAN1,
+	 .flags = IORESOURCE_IRQ,}
+};
+static struct resource flexcan2_resources[] = {
+	{
+	 .start = CAN2_BASE_ADDR,
+	 .end = CAN2_BASE_ADDR + 0x97F,
+	 .flags = IORESOURCE_MEM,},
+	{
+	 .start = MXC_INT_CAN2,
+	 .end = MXC_INT_CAN2,
+	 .flags = IORESOURCE_IRQ,}
+};
+
+static struct platform_device flexcan_devices[] = {
+	{
+	 .name = "FlexCAN",
+	 .id = 0,
+	 .dev = {
+		 .release = mxc_nop_release,
+		 .platform_data = &flexcan_data[0],
+		 },
+	 .num_resources = ARRAY_SIZE(flexcan1_resources),
+	 .resource = flexcan1_resources,},
+	{
+	 .name = "FlexCAN",
+	 .id = 1,
+	 .dev = {
+		 .release = mxc_nop_release,
+		 .platform_data = &flexcan_data[1],
+		 },
+	 .num_resources = ARRAY_SIZE(flexcan2_resources),
+	 .resource = flexcan2_resources,},
+};
+
+static inline void mxc_init_flexcan(void)
+{
+#ifdef CONFIG_FLEXCAN_MXC_SELECT1
+	/* MX25 3stack doesn't use CAN1 */
+	platform_device_register(&flexcan_devices[0]);
+#endif
+	platform_device_register(&flexcan_devices[1]);
+}
+#else
+static inline void mxc_init_flexcan(void)
+{
+}
+#endif
+
 static int __init mxc_init_devices(void)
 {
 	mxc_init_wdt();
@@ -492,6 +551,7 @@ static int __init mxc_init_devices(void)
 	mxc_init_ssi();
 	mxc_init_rtc();
 	imx_init_adc();
+	mxc_init_flexcan();
 
 	return 0;
 }
diff --git a/arch/arm/mach-mx25/mx25_3stack.c b/arch/arm/mach-mx25/mx25_3stack.c
index 1b02605..8c94300 100644
--- a/arch/arm/mach-mx25/mx25_3stack.c
+++ b/arch/arm/mach-mx25/mx25_3stack.c
@@ -428,6 +428,39 @@ static inline void mxc_init_mmc(void)
 }
 #endif
 
+#if defined(CONFIG_CAN_FLEXCAN) || defined(CONFIG_CAN_FLEXCAN_MODULE)
+static void flexcan_xcvr_enable(int id, int en)
+{
+	static int pwdn;
+
+	if (id != 1)		/* MX25 3-stack uses only CAN2 */
+		return;
+
+	if (en) {
+		if (!pwdn++)
+			mxc_set_gpio_dataout(MX25_PIN_D14, 0);
+	} else {
+		if (!--pwdn)
+			mxc_set_gpio_dataout(MX25_PIN_D14, 1);
+	}
+}
+
+struct flexcan_platform_data flexcan_data[] = {
+	{
+	 .core_reg = NULL,
+	 .io_reg = NULL,
+	 .xcvr_enable = flexcan_xcvr_enable,
+	 .active = gpio_can_active,
+	 .inactive = gpio_can_inactive,},
+	{
+	 .core_reg = NULL,
+	 .io_reg = NULL,
+	 .xcvr_enable = flexcan_xcvr_enable,
+	 .active = gpio_can_active,
+	 .inactive = gpio_can_inactive,},
+};
+#endif
+
 /*!
  * Board specific fixup function. It is called by \b setup_arch() in
  * setup.c file very early on during kernel starts. It allows the user to
diff --git a/arch/arm/mach-mx25/mx25_3stack_gpio.c b/arch/arm/mach-mx25/mx25_3stack_gpio.c
index 39eaeac..422bdc5 100644
--- a/arch/arm/mach-mx25/mx25_3stack_gpio.c
+++ b/arch/arm/mach-mx25/mx25_3stack_gpio.c
@@ -483,7 +483,9 @@ void gpio_spi_active(int cspi_mod)
 		mxc_request_iomux(MX25_PIN_CSPI1_SS1, MUX_CONFIG_FUNC);
 		mxc_request_iomux(MX25_PIN_CSPI1_SCLK, MUX_CONFIG_FUNC);
 		mxc_request_iomux(MX25_PIN_CSPI1_RDY, MUX_CONFIG_FUNC);
+#ifndef CONFIG_CAN_FLEXCAN	/* MX25 3-stack uses this pin for CAN2 */
 		mxc_request_iomux(MX25_PIN_GPIO_C, MUX_CONFIG_ALT5); /*SS2*/
+#endif
 		mxc_request_iomux(MX25_PIN_VSTBY_ACK, MUX_CONFIG_ALT2); /*SS3*/
 
 		/* Or if VSTBY_ACK is being used */
@@ -495,7 +497,9 @@ void gpio_spi_active(int cspi_mod)
 		mxc_iomux_set_pad(MX25_PIN_CSPI1_SS1, SPI_PAD_CTL1);
 		mxc_iomux_set_pad(MX25_PIN_CSPI1_SCLK, SPI_PAD_CTL1);
 		mxc_iomux_set_pad(MX25_PIN_CSPI1_RDY, SPI_PAD_CTL1);
+#ifndef CONFIG_CAN_FLEXCAN	/* MX25 3-stack uses this pin for CAN2 */
 		mxc_iomux_set_pad(MX25_PIN_GPIO_C, SPI_PAD_CTL2);
+#endif
 		mxc_iomux_set_pad(MX25_PIN_VSTBY_ACK, SPI_PAD_CTL1);
 
 		mxc_iomux_set_input(MUX_IN_CSPI1_IPP_IND_SS3_B,
@@ -589,7 +593,9 @@ void gpio_spi_inactive(int cspi_mod)
 		mxc_request_gpio(MX25_PIN_CSPI1_SS1);
 		mxc_request_gpio(MX25_PIN_CSPI1_SCLK);
 		mxc_request_gpio(MX25_PIN_CSPI1_RDY);
+#ifndef CONFIG_CAN_FLEXCAN	/* MX25 3-stack uses this pin for CAN2 */
 		mxc_request_gpio(MX25_PIN_GPIO_C); /*SS2*/
+#endif
 		mxc_request_gpio(MX25_PIN_VSTBY_ACK); /*SS3*/
 
 		mxc_free_iomux(MX25_PIN_CSPI1_MOSI, MUX_CONFIG_GPIO);
@@ -598,7 +604,9 @@ void gpio_spi_inactive(int cspi_mod)
 		mxc_free_iomux(MX25_PIN_CSPI1_SS1, MUX_CONFIG_GPIO);
 		mxc_free_iomux(MX25_PIN_CSPI1_SCLK, MUX_CONFIG_GPIO);
 		mxc_free_iomux(MX25_PIN_CSPI1_RDY, MUX_CONFIG_GPIO);
+#ifndef CONFIG_CAN_FLEXCAN	/* MX25 3-stack uses this pin for CAN2 */
 		mxc_free_iomux(MX25_PIN_GPIO_C, MUX_CONFIG_GPIO);
+#endif
 		mxc_free_iomux(MX25_PIN_VSTBY_ACK, MUX_CONFIG_GPIO);
 		break;
 	case 1:
@@ -1151,44 +1159,30 @@ EXPORT_SYMBOL(gpio_inactivate_esai_ports);
  */
 void gpio_can_active(int id)
 {
-#define CAN_PAD_CTL (PAD_CTL_HYS_SCHMITZ | PAD_CTL_PKE_ENABLE | \
-		     PAD_CTL_PUE_PUD | PAD_CTL_100K_PU | PAD_CTL_ODE_OpenDrain)
+#define CAN_PAD_CTL (PAD_CTL_DRV_3_3V | PAD_CTL_PKE_NONE | PAD_CTL_ODE_CMOS | \
+		     PAD_CTL_DRV_NORMAL | PAD_CTL_SRE_SLOW)
+#define CAN_PAD_IN_CTL (PAD_CTL_HYS_CMOS | PAD_CTL_PKE_NONE)
 
 	switch (id) {
 	case 0:
-#if 0
 		/* CAN1 */
 		mxc_request_iomux(MX25_PIN_GPIO_A, MUX_CONFIG_ALT6); /*TXCAN*/
 		mxc_request_iomux(MX25_PIN_GPIO_B, MUX_CONFIG_ALT6); /*RXCAN*/
-#if 0
-		/* Or if FEC is not used */
-		/*TXCAN*/
-		mxc_request_iomux(MX25_PIN_FEC_TX_EN, MUX_CONFIG_ALT4);
-		/*RXCAN*/
-		mxc_request_iomux(MX25_PIN_FEC_RDATA0, MUX_CONFIG_ALT4);
-#endif
 
 		mxc_iomux_set_pad(MX25_PIN_GPIO_A, CAN_PAD_CTL);
-		mxc_iomux_set_pad(MX25_PIN_GPIO_B, CAN_PAD_CTL);
+		mxc_iomux_set_pad(MX25_PIN_GPIO_B, CAN_PAD_IN_CTL);
 
-#endif
+		mxc_iomux_set_input(MUX_IN_CAN1_IPP_IND_CANRX, INPUT_CTL_PATH1);
 		break;
 	case 1:
 		/* CAN2 */
 		mxc_request_iomux(MX25_PIN_GPIO_C, MUX_CONFIG_ALT6); /*TXCAN*/
 		mxc_request_iomux(MX25_PIN_GPIO_D, MUX_CONFIG_ALT6); /*RXCAN*/
 		mxc_request_iomux(MX25_PIN_D14, MUX_CONFIG_ALT5); /*PWDN*/
-#if 0
-		/* Or if FEC is not used */
-		/*TXCAN*/
-		mxc_request_iomux(MX25_PIN_FEC_RDATA1, MUX_CONFIG_ALT4);
-		*RXCAN*/
-		mxc_request_iomux(MX25_PIN_FEC_RX_DV, MUX_CONFIG_ALT4);
-#endif
 
 		mxc_iomux_set_pad(MX25_PIN_GPIO_C, CAN_PAD_CTL);
-		mxc_iomux_set_pad(MX25_PIN_GPIO_D, CAN_PAD_CTL);
-		mxc_iomux_set_pad(MX25_PIN_D14, PAD_CTL_DRV_NORMAL);
+		mxc_iomux_set_pad(MX25_PIN_GPIO_D, CAN_PAD_IN_CTL);
+		mxc_iomux_set_pad(MX25_PIN_D14, CAN_PAD_CTL);
 
 		mxc_iomux_set_input(MUX_IN_CAN2_IPP_IND_CANRX, INPUT_CTL_PATH1);
 
@@ -1211,25 +1205,20 @@ void gpio_can_inactive(int id)
 {
 	switch (id) {
 	case 0:
-#if 0
 		/* CAN1 */
 		mxc_request_gpio(MX25_PIN_GPIO_A); /*TXCAN*/
 		mxc_request_gpio(MX25_PIN_GPIO_B); /*RXCAN*/
 
 		mxc_free_iomux(MX25_PIN_GPIO_A, MUX_CONFIG_FUNC);
 		mxc_free_iomux(MX25_PIN_GPIO_B, MUX_CONFIG_FUNC);
-
-#endif
 		break;
 	case 1:
 		/* CAN2 */
 		mxc_request_gpio(MX25_PIN_GPIO_C); /*TXCAN*/
 		mxc_request_gpio(MX25_PIN_GPIO_D); /*RXCAN*/
-		mxc_request_gpio(MX25_PIN_D14); /*PWDN*/
 
 		mxc_free_iomux(MX25_PIN_GPIO_C, MUX_CONFIG_FUNC);
 		mxc_free_iomux(MX25_PIN_GPIO_D, MUX_CONFIG_FUNC);
-		mxc_free_iomux(MX25_PIN_D14, MUX_CONFIG_FUNC);
 
 		/* Disable input by setting PWDN/TLE6250.INH high */
 		mxc_set_gpio_dataout(MX25_PIN_D14, 1);
diff --git a/include/asm-arm/arch-mxc/mx25.h b/include/asm-arm/arch-mxc/mx25.h
index 13b257a..6402e59 100644
--- a/include/asm-arm/arch-mxc/mx25.h
+++ b/include/asm-arm/arch-mxc/mx25.h
@@ -118,7 +118,7 @@
 #define I2C_BASE_ADDR       (AIPS1_BASE_ADDR + 0x00080000)
 #define I2C3_BASE_ADDR      (AIPS1_BASE_ADDR + 0x00084000)
 #define CAN1_BASE_ADDR      (AIPS1_BASE_ADDR + 0x00088000)
-#define CAN3_BASE_ADDR      (AIPS1_BASE_ADDR + 0x0008C000)
+#define CAN2_BASE_ADDR      (AIPS1_BASE_ADDR + 0x0008C000)
 #define UART1_BASE_ADDR     (AIPS1_BASE_ADDR + 0x00090000)
 #define UART2_BASE_ADDR     (AIPS1_BASE_ADDR + 0x00094000)
 #define I2C2_BASE_ADDR      (AIPS1_BASE_ADDR + 0x00098000)
-- 
1.5.4.4

